### ISSUE TITLE
weird output when using central_longitude != 0

### ISSUE THREAD

**Issue by JosephBARBIERDARNAL** [SciTools/cartopy#2412]
### Description

When using `central_longitude=10`, the output chart becomes unvalid with some horizontal lines (seems to be Canada being deformed in the case below).

- This (seems to) works for any value `centra_longitude` different of 0.
- Same for other projection (I haven't tried them all)

Since the error is a bit too obvious, I doubt I'm first one having the issue, but I haven't found something similar on the internet/issues. Is it an expected behavior?

#### Code to reproduce

This is a copy-pastable example:

```python
import matplotlib.pyplot as plt
import geopandas as gpd
import cartopy.crs as ccrs

proj = ccrs.Mercator(central_longitude=10)

url = "https://raw.githubusercontent.com/holtzy/The-Python-Graph-Gallery/master/static/data/all_world.geojson"
world = gpd.read_file(url)
world = world[~world['name'].isin(["Antarctica", "Greenland"])]
world = world.to_crs(proj.proj4_init)

fig, ax = plt.subplots(figsize=(12, 8), dpi=300, subplot_kw={'projection':proj})
ax.set_axis_off()
world.boundary.plot(ax=ax)
plt.show()
```

<img width="1108" alt="Screenshot 2024-07-03 at 15 30 42" src="https://github.com/SciTools/cartopy/assets/79746670/0cade372-b95c-4902-a4e9-35c334f0ff5a">

#### Traceback

None

### Operating system

MacOS Sonoma

### Cartopy version

0.23.0

**Comment by lgolston** [comment]
Yes, I think it is expected (see note in geopandas docs about objects crossing the dateline: https://geopandas.org/en/stable/docs/reference/api/geopandas.GeoDataFrame.to_crs.html)

Here is a different version that should be working:
```
proj = ccrs.Mercator(central_longitude=10)

url = "https://raw.githubusercontent.com/holtzy/The-Python-Graph-Gallery/master/static/data/all_world.geojson"
world = gpd.read_file(url)
world = world[~world['name'].isin(["Antarctica", "Greenland"])]

fig, ax = plt.subplots(figsize=(12, 8), dpi=300, subplot_kw={'projection':proj})
ax.add_geometries(world["geometry"], crs=ccrs.PlateCarree(), facecolor='none')
plt.show()
```

**Comment by schlunma** [comment]
I have a similar problem with `contourf` and the `Robinson` projection:

```py
import iris
import iris.plot
import matplotlib.pyplot as plt
import cartopy.crs as ccrs

cube = iris.load_cube('test_central_lon.nc')

fig = plt.figure()
fig.add_subplot(projection=ccrs.Robinson())
iris.plot.contourf(cube, cmap='bwr')
```

gives a perfectly reasonable plot:

![grafik](https://github.com/user-attachments/assets/2ebb68a3-33ab-4913-8573-274685cd3a5f)

Running the same script with `ccrs.Robinson(central_longitude=10.0)` gives:

![grafik](https://github.com/user-attachments/assets/755fa5d8-ea51-46f4-bff9-194ee1cfa490)


This is the file, hope that works fine to share that via GitHub (just remove the `.txt` extension)

[test_central_lon.nc.txt](https://github.com/user-attachments/files/17478494/test_central_lon.nc.txt)

**Comment by schlunma** [comment]
I opened a new [issue](https://github.com/SciTools/cartopy/issues/2457) for the problem described above; this also happens for `central_longitude=0.0`.
